packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,306 @@
+    typedef typename Adapter::part_t part_t;
+    typedef typename Adapter::user_t user_t;
+    typedef typename Adapter::userCoord_t userCoord_t;
+    typedef Tpetra::Map<lno_t, gno_t> map_t;
+    
+    RCP<const Environment> env = rcp(new Environment(problemComm));
+    RCP<const base_adapter_t> b_adapter = adapter;
+    modelFlag_t flags;
+    flags.reset();
+    
+    flags.set(REMOVE_SELF_EDGES);
+  
+    env->debug(DETAILED_STATUS, "	building graph model");
+    //const RCP<const GraphAdapter<user_t>> g_adapter = adapter;
+    RCP<const GraphModel<base_adapter_t>> model = rcp(new GraphModel<base_adapter_t>(b_adapter,env,
+			                                       problemComm, flags));
+    env->debug(DETAILED_STATUS, "	graph model built");
+    
+    //Get vertex GIDs, in a locally indexed array
+    ArrayView<const gno_t> vtxIDs;
+    ArrayView<StridedData<lno_t, scalar_t> > vwgts;
+    size_t nVtx = model->getVertexList(vtxIDs, vwgts);
+    //at this point, weights are not used.
+ 
+    //Get the edge information
+    ArrayView<const gno_t> adjs; 
+    ArrayView<const offset_t> offsets;
+    ArrayView<StridedData<lno_t, scalar_t> > ewgts;
+    size_t nEdge = model->getEdgeList(adjs, offsets, ewgts);
+    //edge weights are not used either.
+  
+    //copy data over to regular arrays
+    gno_t* out_edges = NULL;
+    typename map_t::local_ordinal_type* out_offsets = NULL;
+    gno_t* global_ids = NULL;
+    TPL_Traits<gno_t, const gno_t>::ASSIGN_ARRAY(&out_edges, adjs);
+    TPL_Traits<typename map_t::local_ordinal_type, const offset_t>::ASSIGN_ARRAY(&out_offsets, offsets);
+    TPL_Traits<gno_t, const gno_t>::ASSIGN_ARRAY(&global_ids, vtxIDs);
+  
+    //get the rank of the current process
+    int me = problemComm->getRank();
+  
+    //create the Tpetra map for the global indices
+    Tpetra::global_size_t dummy = Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid();
+    RCP<const map_t > map = rcp(new map_t(dummy, vtxIDs, 0, problemComm)); 
+  
+    //sentinel value to detect unsuccessful searches
+    lno_t fail = Teuchos::OrdinalTraits<lno_t>::invalid();
+
+    //map to make sure ghosted vertices aren't duplicated in the Tpetra Maps
+    std::unordered_map<gno_t,lno_t> ghost_map;
+    //The new local IDs for each ghosted vertex
+    std::vector<lno_t> ghostGIDs;
+    int nGhosts = 0;
+    //count how many ghosts are in the edge list, to create the mapWithCopies.
+    for(size_t i = 0; i < nEdge; i++){
+      if(map->getLocalElement(adjs[i]) == fail){
+        if(ghost_map.count(adjs[i]) == 0){
+          ghost_map[adjs[i]] = nVtx + nGhosts;
+          ghostGIDs.push_back(adjs[i]);
+          nGhosts++;
+        }
+      }
+    }
+  
+    //build nVtx + nGhosts size array of grounding flags
+    bool* grounding = new bool[nVtx+nGhosts];
+    for(size_t i = 0; i < nVtx+nGhosts; i++){
+      if(i < nVtx){
+        grounding[i] = basalFriction[i];
+      }
+      else grounding[i] = false;
```

Comment:
Is it necessary for grounding to store false values for the ghosts?  
If we don't need the grounding info for the ghosts, we could avoid the entire copy from basalFriction to grounding.
See above note in Zoltan2_IceProp.hpp.

Reply:
I do not think this is necessary. I have reduced the grounding size to just the locally owned vertices.